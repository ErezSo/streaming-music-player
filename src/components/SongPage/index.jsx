import React from "react";

const styles = {
  listItem: {
    position: "relative",
    height: "13em"
  },
  imageContainer: {
    float: "left"
  },
  image: {
    height: "150px",
    width: "150px",
    position: "absolute"
  },
  button: {
    fontSize: "2em",
    borderRadius: "50%",
    position: "absolute",
    top: "3.5em",
    left: "3.8em"
  },
  deatilsList: {
    position: "absolute",
    left: "12em",
    listStyleType: "none"
  }
};

const song = { id: "", title: "", user: { username: "" } };

function findSong(artists, id) {
  return artists.length > 0
    ? artists.find(artist => artist.id === id, 10)
    : song;
}

const convertToYear = date => new Date(date).getFullYear();

// Source: https://stackoverflow.com/a/37096512/4640352
function secondsToHms(d) {
  d = Number(d);
  var h = Math.floor(d / 3600);
  var m = Math.floor((d % 3600) / 60);
  var s = Math.floor((d % 3600) % 60);

  var hDisplay = h > 0 ? `${h}:` : "";
  var mDisplay = m > 0 ? `${m}:` : "";
  var sDisplay = s > 0 ? s : "";
  return hDisplay + mDisplay + sDisplay;
}

function SongPage({
  artists,
  match: {
    match: {
      params: { id }
    }
  },
  playSong
}) {
  const song = findSong(artists, id);
  return (
    <div className="panel panel-default">
      <div className="panel-heading">
        <strong>
          {song.user.username} - {song.title}
        </strong>
      </div>
      <ul className="list-group">
        <li className="list-group-item" style={styles.listItem}>
          <div style={styles.imageContainer}>
            <img
              src={song.artwork_url}
              alt="song"
              className="img-rounded"
              style={styles.image}
            />
            <button
              style={styles.button}
              className="btn btn-warning glyphicon glyphicon-play-circle"
              onClick={() => playSong(song.stream_url)}
            />
          </div>
          <div>
            <ul style={styles.deatilsList}>
              <li>Genre: {song.genre}</li>
              <li>Year released: {convertToYear(song.release_date)}</li>
              <li>Song duration {secondsToHms(song.duration)}</li>
            </ul>
          </div>
        </li>
      </ul>
    </div>
  );
}

export default SongPage;
