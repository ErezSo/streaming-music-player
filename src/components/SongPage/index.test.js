import React from "react";
import { shallow } from "enzyme";
import SongPage from "./";

describe("SongPage", () => {
  let wrapper;
  const spyOnPlaySong = jest.fn();
  const props = {
    artists: [
      {
        id: "43",
        release_date: "2018-05-14 09:46:01",
        downloadable: "0",
        title: "Bestami Turna - Soundpark 14.04.2018",
        artwork_url:
          "https://images.hearthis.at/1/5/2/_/uploads/283048/image_track/1914537/w500_q70_----1525319281387.jpg",
        user: {
          id: "283048",
          username: "TDSmix"
        },
        stream_url:
          "https://hearthis.at/tdsmix/bestami-turna-soundpark-14042018/listen/?s=rCW"
      }
    ],
    match: {
      match: {
        params: {
          id: "43"
        }
      }
    },
    playSong: spyOnPlaySong
  };

  beforeEach(() => {
    wrapper = shallow(<SongPage {...props} />);
  });

  it("renders", () => {
    expect(wrapper).toHaveLength(1);
  });

  it("renders correctly", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('plays the selected song when clicking on "play"', () => {
    wrapper.find(".glyphicon-play-circle").simulate("click");
    expect(spyOnPlaySong).toHaveBeenCalledTimes(1);
  });
});
