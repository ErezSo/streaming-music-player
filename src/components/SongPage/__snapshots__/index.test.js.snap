// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongPage renders correctly 1`] = `
"<div className=\\"panel panel-default\\">
  <div className=\\"panel-heading\\">
    <strong>
      TDSmix
       - 
      Bestami Turna - Soundpark 14.04.2018
    </strong>
  </div>
  <ul className=\\"list-group\\">
    <li className=\\"list-group-item\\" style={{...}}>
      <div style={{...}}>
        <img src=\\"https://images.hearthis.at/1/5/2/_/uploads/283048/image_track/1914537/w500_q70_----1525319281387.jpg\\" alt=\\"song\\" className=\\"img-rounded\\" style={{...}} />
        <button style={{...}} className=\\"btn btn-warning glyphicon glyphicon-play-circle\\" onClick={[Function: mockConstructor]} />
      </div>
      <div>
        <ul style={{...}}>
          <li>
            Genre: 
          </li>
          <li>
            Year released: 
            2018
          </li>
          <li>
            Download:
             
            <a href={[undefined]} className=\\"glyphicon glyphicon-circle-arrow-down\\" alt=\\"download song\\" download={true} style={{...}} />
          </li>
        </ul>
      </div>
    </li>
  </ul>
</div>"
`;
